### PR TITLE
Update Infinispan version to 14.0.2.Final

### LINKED ISSUE #12494: Update Infinispan version to 14.x.x
After this change it is possible to start with the following issues: 
- #12493
- #10988
- #12497

### PR DESCRIPTION
**Pull request by martin-kanis**
Closes #12494

Updating Infinispan version to 14.0.0.Final. We should use 14.0.1.Final version once it is released.

### CODE REVIEW

**Comment by martin-kanis** [comment]
@pedroigor Can you please clarify purpose of `ClusterConfigDistTest.testCustomCacheStackInConfigFile` test? After Infinispan update, I had to fix jgroups configuration in `cache-ispn-asym-enc.xml` and server now starts without any error. Without any change in `cache-ispn-asym-enc.xml` it complains about `stack.combine` and `stack.position` properties.

**Comment by pedroigor** [comment]
@martin-kanis The purpose is to make sure the `stack` defined in the configuration is recognized. If the new version complains about those two attributes, it should be a matter of updates to the configuration schema. Your changes to that file should be fine as the test expects the server to fail to start due to a missing keystore meaning the custom stack was recognized.

**Comment by martin-kanis** [comment]
> @martin-kanis The purpose is to make sure the `stack` defined in the configuration is recognized. If the new version complains about those two attributes, it should be a matter of updates to the configuration schema. Your changes to that file should be fine as the test expects the server to fail to start due to a missing keystore meaning the custom stack was recognized.
> 
> Perhaps we should mention this upgrade, if not done already, in the upgrade/migration guide.

I see. Thanks for the clarification. For some reason, server is not complaining about not existing keystore anymore. Instead of expecting an error, I check if the custom jgroups stack was picked from the configuration.

**Comment by martin-kanis** [comment]
@mhajas Maybe we can proceed with version 14.0.0.Final now, start working on the dependent issues and upgrade to 14.0.1.Final when available. WDYT?

**Comment by martin-kanis** [comment]
> @martin-kanis Sounds ok to me, when is 14.0.1 expected to be out?

I don't know that. But I can see 14.0.1.Final tag on Infinispan's GH already. Maybe it will be out before this PR is ready to be merged :)

**Comment by martin-kanis** [comment]
I did some more changes to our default udp config file to match Infinispan's defaults https://github.com/infinispan/infinispan/blob/main/core/src/main/resources/default-configs/default-jgroups-udp.xml
Lets see if everything passes.

**Comment by martin-kanis** [comment]
\rerun

File: model/infinispan/src/main/resources/default-configs/default-keycloak-jgroups-udp.xml
Comment (mhajas) [inline model/infinispan/src/main/resources/default-configs/default-keycloak-jgroups-udp.xml:50]: @hmlnarik Do you know why default `jgroups.join_timeout` was set to 500 and default in Infinispan [default config was 2000](https://github.com/infinispan/infinispan/blob/11.0.9.Final/core/src/main/resources/default-configs/default-jgroups-udp.xml#L52)? Was it no purpose or we want to use default ISPN value?

Comment (hmlnarik) [reply model/infinispan/src/main/resources/default-configs/default-keycloak-jgroups-udp.xml:50]: TL;DR: It was on purpose: The lower the time, the faster the startup.

The join timeout determines time to wait for nodes to respond to the discovery request. In the usual settings, it is run on a single node / virtual network with virtually zero network latency so the discovery finishes in almost no time.

Comment (mhajas) [reply model/infinispan/src/main/resources/default-configs/default-keycloak-jgroups-udp.xml:50]: I see. @martin-kanis Can you please change it to 500 then?

Also, @hmlnarik Could you please look at the changes in this PR whether you can see some other option that needs to be set differently than the default? We went through the changes with @martin-kanis and it seemed all others were the default, but we might have missed something.

**Review by mhajas — APPROVED** [review]
Thanks for the PR @martin-kanis. The changes look good to me!

The failure in the Model tests seems irrelevant as it also failed in [this nightly build](https://github.com/keycloak/keycloak/actions/runs/3305238747/jobs/5455247481).

There was also Jenkins running on this PR and it seems there are no additional failures caused by this:
https://master-jenkins.redhat.com/job/universal-test-pipeline-server/2310/

We need to wait for `quarkus-map` `group1` to pass and if it passes I will merge.

File: model/infinispan/src/main/resources/default-configs/default-keycloak-jgroups-udp.xml
Comment (hmlnarik) [reply model/infinispan/src/main/resources/default-configs/default-keycloak-jgroups-udp.xml]: LGTM as far as I can tell

**Review by abstractj — COMMENTED** [review]
@martin-kanis @mhajas I'm supportive to always keep our dependencies up to date, the changes proposed look good to me. Although, we need to stay in sync with Quarkus releases. At the moment Quarkus 2.13.3 is still using Infinispan 13.0.11. Please see:
https://github.com/quarkusio/quarkus/blob/2.13.3.Final/bom/application/pom.xml#L133

I'd say that we may need to put those changes on hold, until the next Quarkus release. cc @stianst @hmlnarik

**Comment by hmlnarik** [comment]
> @martin-kanis @mhajas I'm supportive to always keep our dependencies up to date, the changes proposed look good to me. Although, we need to stay in sync with Quarkus releases. At the moment Quarkus 2.13.3 is still using Infinispan 13.0.11. Please see: https://github.com/quarkusio/quarkus/blob/2.13.3.Final/bom/application/pom.xml#L133
> 
> I'd say that we may need to put those changes on hold, until the next Quarkus release. cc @stianst @hmlnarik

Thank you for checking this one.
I agree these changes should be held back until Keycloak 20 is released. After that I am in favour of including this PR, since Infinispan 14 upgrade is a necessary condition for several issues, and it is already in [quarkus 2.14 branch](https://github.com/quarkusio/quarkus/blob/2.14/bom/application/pom.xml#L133) which we need to upgrade to anyway.
WDYT?

**Comment by abstractj** [comment]
Makes sense @hmlnarik, but I'd suggest upgrading Quarkus first before merging it. If we do in the opposite order, the chances of having dependencies out of sync can be high.

**Comment by hmlnarik** [comment]
@pedroigor Would you know ETA of update to Quarkus 2.14?

**Review by mhajas — CHANGES_REQUESTED** [review]
Removing my approval so we don't merge this accidentally.

**Comment by abstractj** [comment]
@hmlnarik 2.14.0.CR1 was tagged 7 hours ago https://github.com/quarkusio/quarkus/tags. Reviewing their milestones, I can't see 2.14.0 release yet. As soon as their team releases, if we don't find any breaking changes, it should not be a problem to update those dependencies.

**Comment by pedroigor** [comment]
@martin-kanis Can we upgrade to 14.0.2.Final instead?

FYI, the upgrade should also help when upgrading to Quarkus 2.14+. See https://github.com/quarkusio/quarkus/issues/28893.

**Comment by martin-kanis** [comment]
> @martin-kanis Can we upgrade to 14.0.2.Final instead?
> 
> FYI, the upgrade should also help when upgrading to Quarkus 2.14+. See [quarkusio/quarkus#28893](https://github.com/quarkusio/quarkus/issues/28893).

Sure, 14.0.2.Final is fine. I can update the PR, so it's eventually ready when we upgrade Quarkus.

**Comment by martin-kanis** [comment]
> I was wondering whether the name of the file [default-keycloak-jgroups-udp.xml](https://github.com/keycloak/keycloak/pull/14925#diff-eb246d3cde4879eae01bd89aa0bfcb7919d7df5533c81293f11c2094dfb7647f) is actually correct. I have seen it is referenced at
> 
> https://github.com/keycloak/keycloak/blob/f6985949b65da2dfb527e7127fad3d31f0dbba60/model/infinispan/src/main/java/org/keycloak/connections/infinispan/DefaultInfinispanConnectionProviderFactory.java#L212
> 
> . However, there is an indirect reference to a similar default file at
> https://github.com/keycloak/keycloak/blob/f6985949b65da2dfb527e7127fad3d31f0dbba60/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java#L124
> 
> and that would indicate the name should be `default-jgroups-udp.xml`. Because this file will be looked up when --cache-stack=UDP is actually passed as a parameter. But I might be wrong, I was just confused.

Maybe you are right. `default-keycloak-jgroups-udp.xml` filename is from older date than https://github.com/keycloak/keycloak/blob/f6985949b65da2dfb527e7127fad3d31f0dbba60/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java#L124 so I would investigate whether this piece of code is really targeting `default-configs/default-keycloak-jgroups-udp.xml` file. I so, there is a bug in `configureTransportStack()` @pedroigor Can you check it please?

**Comment by pedroigor** [comment]
> so I would investigate whether this piece of code is really targeting `default-configs/default-keycloak-jgroups-udp.xml` file. If so, there is a bug in `configureTransportStack()`. @pedroigor Can you check it please?

It is. We try to map a stack to the corresponding file from Infinispan JAR.

**Comment by martin-kanis** [comment]
> > so I would investigate whether this piece of code is really targeting `default-configs/default-keycloak-jgroups-udp.xml` file. If so, there is a bug in `configureTransportStack()`. @pedroigor Can you check it please?
> 
> It is. We try to map a stack to the corresponding file from Infinispan JAR.

Thanks for confirmation. Maybe we could reopen https://github.com/keycloak/keycloak/issues/10341 (or create a new issue) as there is the error mentioned related exactly to this. 
```
2022-02-20 19:42:40,402 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: org.infinispan.commons.CacheConfigurationException: ISPN000085: Error while trying to create a channel using the specified configuration file: default-configs/default-jgroups-kubernetes.xml
2022-02-20 19:42:40,403 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: ISPN000085: Error while trying to create a channel using the specified configuration file: default-configs/default-jgroups-kubernetes.xml
```
For `tcp` and `kubernetes` there is no default file in `default-configs`. For `udp` the filename is not matching.

**Review by mhajas — APPROVED** [review]
Thanks for the PR @martin-kanis!

**Pull request merged**
